### PR TITLE
support kernel chaos

### PR DESCRIPTION
**Pull request by ethercflow**
### What problem does this PR solve? 

Support kernel fault injection, please read [inject_example.txt](https://github.com/iovisor/bcc/blob/master/tools/inject_example.txt) to learn more.

### What is changed and how does it work?

1. Add a new chaos type;
2. Send request to bpfki instead of chaos-daemon

### Check List 

Tests 

 - No code

Code changes

 - Has Go code change

Side effects

 - None

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: api/v1alpha1/kernelchaos_types.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	return in.Value
+}
+
+type FailKernRequest struct {
+	FailType int32 `json:"failtype"`
+	Headers []string `json:"headers"`
+	Callchain []Frame `json:"callchain"`
+	Probability uint32 `json:"probability"`
+	Times uint32 `json:"times"`
+};
```
Comment (you06) [inline api/v1alpha1/kernelchaos_types.go]: May we remove the semicolon?

File: api/v1alpha1/kernelchaos_types.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+type Frame struct {
+	Funcname string `json:"funcname"`
+	Parameters string `json:"parameters"`
+	Predicate string `json:"predicate"`
+};
```
Comment (you06) [inline api/v1alpha1/kernelchaos_types.go]: ditto

**Comment by Yisaer** [comment]
@ethercflow  Can this request split into several requests? I think we can start from the API definition.

File: api/v1alpha1/kernelchaos_types.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	// If `FixedPercentPodMod`, provide a number from 0-100 to specify the max % of pods the server can do chaos action.
+	// If `RandomMaxPercentPodMod`,  provide a number from 0-100 to specify the % of pods to do chaos action
+	// +optional
+	Value string `json:"value"`
+
+	// Selector is used to select pods that are used to inject chaos action.
+	Selector SelectorSpec `json:"selector"`
+
+	// FailKernRequest defines the request of kernel injection
+	FailKernRequest FailKernRequest `json:"failKernRequest"`
```
Comment (AngleNet) [inline api/v1alpha1/kernelchaos_types.go]: s/FailKernRequest/FailKernelRequest?

File: pkg/utils/grpc.go [hunk 4]
```diff
@@ -32,12 +32,7 @@ const DefaultRPCTimeout = 60 * time.Second
 // RPCTimeout specifies timeout of RPC between controller and chaos-operator
 var RPCTimeout = DefaultRPCTimeout
 
-func CreateGrpcConnection(ctx context.Context, c client.Client, pod *v1.Pod) (*grpc.ClientConn, error) {
-	port := os.Getenv("CHAOS_DAEMON_PORT")
-	if port == "" {
-		port = "8080"
-	}
-
+func CreateGrpcConnection(ctx context.Context, c client.Client, pod *v1.Pod, port string) (*grpc.ClientConn, error) {
```
[the remark this answers is not in the document]
Comment (AngleNet) [reply pkg/utils/grpc.go]: Adding port to the signature seems to be redudant ?

Comment (yeya24) [reply pkg/utils/grpc.go]: I think it's fine. But each time, get the port from env variables like `CreateGrpcConnection(ctx, r.Client, pod, os.Getenv("CHAOS_DAEMON_PORT"))` is not good. It is better to use variable keep the port (daemon/bpfki). But it is fine in this PR. Maybe we can open an issue about it and change it in the future.

File: controllers/kernelchaos/types.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	"github.com/pingcap/chaos-mesh/controllers/twophase"
+	"github.com/pingcap/chaos-mesh/pkg/utils"
+
+	pb "github.com/pingcap/chaos-mesh/pkg/chaosdaemon/pb"
+	pb_ "github.com/pingcap/chaos-mesh/pkg/chaoskernel/pb"
+)
+
+const kernelChaosMsg = "kernel is injected with %v"
+
+// Reconciler is KernelChoas reconciler
```
Comment (yeya24) [inline controllers/kernelchaos/types.go]: s/KernelChoas/KernelChaos/g

File: controllers/kernelchaos/types.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+			Message:   fmt.Sprintf(kernelChaosMsg, kernelchaos.Spec.FailKernRequest),
+		}
+
+		kernelchaos.Status.Experiment.Pods = append(kernelchaos.Status.Experiment.Pods, ps)
+	}
+
+	return nil
+}
+
+// Recover means the reonciler recover the chaos action
```
Comment (yeya24) [inline controllers/kernelchaos/types.go]: This should be `Recover means the reconciler recovers the chaos action`

File: helm/chaos-mesh/values.yaml [hunk 7]
```diff
@@ -139,3 +139,10 @@ preJobs:
 
 postJobs:
   podAnnotations: {}
+
+bpfki:
+  image: pingcap/chaos-kernel:latest
+  imagePullPolicy: Always
+  grpcPort: 50051
+
+  podAnnotations: {}
```
Comment (yeya24) [inline helm/chaos-mesh/values.yaml]: Seems bpfki is deployed in the same pod with chaos-daemon, so what about moving this part to be part of `chaos-daemon` in `values.yaml`? What do you think?

And I am not that familar with bpfki, can I ask is there any case that we don't want to deploy bpfki, only deploy chaos-daemon?  If that, I think it is better to add an `enable` option

Comment (zhouqiang-cl) [reply helm/chaos-mesh/values.yaml]: bpfki have performance issues on every container(not only the inject one)

File: pkg/chaoskernel/pb/bpfki.proto [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+service BPFKIService {
+  rpc SetTimeVal(BumpTimeRequest) returns (StatusResponse) {}
+  rpc RecoverTimeVal(BumpTimeRequest) returns (StatusResponse) {}
+  rpc SetTimeSpec(BumpTimeRequest) returns (StatusResponse) {}
+  rpc RecoverTimeSpec(BumpTimeRequest) returns (StatusResponse) {}
+  rpc FailMMOrBIO(FailKernRequest) returns (StatusResponse) {}
+  rpc RecoverMMOrBIO(FailKernRequest) returns (StatusResponse) {}
+  rpc FailSyscall(FailSyscallRequest) returns (StatusResponse) {}
```
Comment (cwen0) [inline pkg/chaoskernel/pb/bpfki.proto]: Where does use `FailSyscall`?

File: doc/kernel_chaos.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+> Cannot allocate memory
+> Invalid argument
+
+> For more sample files, see [examples](https://github.com/chaos-mesh/bpfki/tree/develop/examples). You can edit them as needed. 
+
+Description:
+
+* **mode** defines the mode to select pods.
+* **selector** specifies the target pods for chaos injection.
+* **failkernRequest** defines the specified injection mode (kmalloc,bio,etc) with a call chain and an optional set of predicates.
```
Comment (cwen0) [inline doc/kernel_chaos.md]: add more description about failKernRequest

Comment (cwen0) [inline doc/kernel_chaos.md]: What is the difference between `kmalloc` and `bio`?

Comment (ethercflow) [reply doc/kernel_chaos.md]: `bio` is a concept of block layer, so it's used to inject io request error. 
`kmalloc` is the mostly caller of `slab`.

**Comment by cwen0** [comment]
@ethercflow fix conflict?

File: api/v1alpha1/kernelchaos_types.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	// Probability indicates the fails with probability.
+	// If you want 1%, please set this field with 1.
+	Probability uint32 `json:"probability,omitempty"`
+
+	// Times indicates the max times of fails.
+	Times uint32 `json:"times,omitempty"`
+}
+
+// Frame defines the function signature and predicate in function's body
+type Frame struct {
```
Comment (cwen0) [inline api/v1alpha1/kernelchaos_types.go]: add some comments for these

File: controllers/kernelchaos_controller.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+	reconciler := kernelchaos.Reconciler{
+		Client: r.Client,
+		Log:    logger,
+	}
+
+	return reconciler.Reconcile(req)
+}
+
+// SetupWithManager setups a kernel chaos reconciler on controller-manager
```
Comment (fewdan) [inline controllers/kernelchaos_controller.go]: “setup” is not a verb.   Maybe we should s/setups/sets up/g ?

**Pull request merged**